metaflow/task.py

```diff
@@ -513,204 +519,239 @@ def run_step(
+                    deco.task_post_step(
+                        step_name,
+                        self.flow,
+                        self.flow._graph,
+                        retry_count,
+                        max_user_code_retries,
+                    )
 
-            for deco in decorators:
-                deco.task_post_step(
-                    step_name,
-                    self.flow,
-                    self.flow._graph,
-                    retry_count,
-                    max_user_code_retries,
-                )
+                self.flow._task_ok = True
+                self.flow._success = True
 
-            self.flow._task_ok = True
-            self.flow._success = True
-
-        except Exception as ex:
-            tsk_msg = {
-                "task_id": task_id,
-                "exception_msg": str(ex),
-                "msg": "task failed with exception",
-                "step_name": step_name,
-                "run_id": run_id,
-                "flow_name": self.flow.name,
-            }
-            logger.log(tsk_msg)
+            except Exception as ex:
+                with self.monitor.count("metaflow.task.exception"):
+                    # Log both the metric and the exception message
+                    self.event_logger.log(
+                        {
+                            "event_name": "metaflow.task.exception",
+                            "event_value": 1,
+                            "log_type": "ERROR",
+                            "log_value": str(ex),
+                            "traceback": traceback.format_exc(),
+                        }
+                    )
+                    pass
+
+                exception_handled = False
+                for deco in decorators:
+                    res = deco.task_exception(
+                        ex,
+                        step_name,
+                        self.flow,
+                        self.flow._graph,
+                        retry_count,
+                        max_user_code_retries,
+                    )
+                    exception_handled = bool(res) or exception_handled
+
+                if exception_handled:
+                    self.flow._task_ok = True
+                else:
+                    self.flow._task_ok = False
+                    self.flow._exception = MetaflowExceptionWrapper(ex)
+                    print("%s failed:" % self.flow, file=sys.stderr)
+                    raise
+
+            finally:
+                if self.ubf_context == UBF_CONTROL:
+                    self._finalize_control_task()
+
+                # Emit metrics to logger/monitor sidecar implementations
+                with self.monitor.count("metaflow.task.end"):
+                    self.event_logger.log(
+                        {
+                            "event_name": "metaflow.task.end",
+                            "event_value": 1,
+                        }
+                    )
+                    pass
```

Comment:
same here - why is a `pass` needed? is it to signify end of an indent block or something more? if the former, we can avoid introducing pass to maintain consistency of style in the code base

Reply:
Will remove in updated PR